api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+	pb "github.com/textileio/textile/v2/api/sendfild/pb"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	collectionName = "sendfil"
+)
+
+var _ pb.SendFilServiceServer = (*Service)(nil)
+
+var log = logging.Logger("sendfil")
+
+type msgCid struct {
+	Cid       string    `bson:"cid"`
+	CreatedAt time.Time `bson:"created_at"`
+}
+
+type txn struct {
+	ID            primitive.ObjectID `bson:"_id"`
+	From          string             `bson:"from"`
+	To            string             `bson:"to"`
+	AmountNanoFil int64              `bson:"amount_nano_fil"`
+	MessageCids   []msgCid           `bson:"message_cids"`
+	MessageState  pb.MessageState    `bson:"message_state"`
+	Waiting       bool               `bson:"waiting"`
+	FailureMsg    string             `bson:"failure_msg"`
+	CreatedAt     time.Time          `bson:"created_at"`
+	UpdatedAt     time.Time          `bson:"updated_at"`
+}
+
+func (t txn) latestMsgCid() (msgCid, error) {
+	if len(t.MessageCids) == 0 {
+		return msgCid{}, fmt.Errorf("no message cids found")
+	}
+	return t.MessageCids[len(t.MessageCids)-1], nil
+}
+
+type Service struct {
+	clientBuilder lotus.ClientBuilder
+	col           *mongo.Collection
+	server        *grpc.Server
+	waiting       map[primitive.ObjectID]chan waitResult
+	config        Config
+	ticker        *time.Ticker
+	waitingLck    sync.Mutex
+}
+
+type Config struct {
+	Listener           net.Listener
+	ClientBuilder      lotus.ClientBuilder
+	MongoUri           string
+	MongoDbName        string
+	MessageWaitTimeout time.Duration
+	MessageConfidence  uint64
+	RetryWaitFrequency time.Duration
+	Debug              bool
+}
+
+func New(ctx context.Context, config Config) (*Service, error) {
+	if config.Debug {
+		if err := util.SetLogLevels(map[string]logging.LogLevel{
+			"sendfil": logging.LevelDebug,
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(config.MongoUri))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to mongo: %v", err)
+	}
+	db := client.Database(config.MongoDbName)
+	col := db.Collection(collectionName)
+	if _, err := col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys: bson.D{primitive.E{Key: "from", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "to", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "amount_nano_fil", Value: 1}},
+		},
+		// MongoDB automatically creates a multikey index if any indexed field is an array;
+		// you do not need to explicitly specify the multikey type.
+		// https://docs.mongodb.com/manual/core/index-multikey/
+		{
+			Keys:    bson.D{primitive.E{Key: "message_cids.cid", Value: 1}},
+			Options: options.Index().SetUnique(true),
+		},
```

Comment:
Cool mongo feature where you can index values nested in a list, in this case, uniquely.